README.md
### Cinema 4D渲染工作流程插件

<aside>
🔒 当前版本 ：1012

</aside>

文档移植中，源文档可以查看 ： [RenderFlow](https://dunhou.gitbook.io/dh-renderflow/)

# Download--下载

[BOGHMA 社区下载](https://community.boghma.com/)

[Gitee（国内源）](https://gitee.com/DunHouGo/c4dplugin_RenderFlow/repository/archive/master.zip)

[Github（国外源）](https://github.com/DunHouGo/c4dplugin_RenderFlow/archive/refs/heads/master.zip)

# Install--安装

- 推荐使用PluginManager统一管理安装更新：[PluginManager](https://www.notion.so/Plugin-Manager-72c5fe979541467187af2060fe330e80)
- 解压文件夹,复制到C4D的plugin文件夹下

> `%AppData%\Maxon\Maxon Cinema 4D R2x_xxxxxxxx\plugins\CINEMA 4D R2x\plugins\Boghma`
> 

<aside>
⚠️ 不要更改文件夹结构和名称，会导致插件报错
  
</aside>

# Main Function--主要功能

- 根据当前选择渲染引擎切换对应的优化功能
- 智能材质编辑器：根据所选对象或材质激活对应的材质编辑器
- 根据所选对象智能创建材质 、灯光、 标签
- 统一的流程工具
- 集成的管理器：包括WIP
- 根据渲染引擎的智能散布
- 建模相关小工具
- MG相关小工具
- 标签相关小工具
- 材质相关小工具

# How to use--使用

- 正确安装插件
- 默认快捷键为 ~ 飘号键(Tab上方键)
- 如果快捷键失效，切换中英文输入法
- 仍然无效，检查是否安装正常，检查是否有快捷键冲突(Shift + F12查看)
- 仍然无效，按Shift + F10，唤出Console控制台查看Python部分报错

# Support Engine--支持渲染器

- Octane
- Redshift
- Arnold

# Shortcut--快捷键

- 默认快捷键为 ` ( Tab键上方 )
- 如果不起作用 , 尝试切换英文输入法 , 并且检查是否正确安装
- 快捷键可以按照个人喜好更改( Shift+F12 ) , 更改后默认快捷键失效
- 推荐使用Add添加自定义快捷键 , 而不是Assign

# Video --视频教程

暂无
### 由于使用NodeAssetInterface api ，R25以下版本会报错, 暂时没有做低版本兼容


ChangeLog.md
###  ==========  UPDATE LIST  ==========  ###
# For RenderFlow

# 1000
- 旧版脚本 ：https://community.boghma.com/d/22
- RenderFlow插件化，更完善和封闭
- 全新界面，支持Octane，Redshift，Arnold三大主流渲染器
- 统一默认Standard Surface，迁移到Node Editor工作流程，弃用传统xpresso界面
- 新增Basic Flow，增加通用功能
 
# 1001
- 2022.08.20 fix
- 增加不自动指认快捷键版本
- 全面支持撤销
- 增加细化的使用文档和gif示意图
- Scatter 显示bug ==》 调用BaseDocument.SetActiveObject解决

# 1002
- 自动添加GSG link标签
- 目标灯光现在会自动指认一张灯光贴图和HDR link （Arnold自动指认tx）
- HDR（Dome Light）在会自动指认一张HDR和HDR link （Arnold自动指认tx）
- HDR link默认不指认任何贴图，保持工程整洁
- 新增合并材质，新建材质替换所有选择材质的指认，Alt可删除原材质
- 优化RS proxy和BakeSet流程
- 
# 1003
- 2022.12.03 fix
- 更新文件夹结构，适配标准插件结构
- 更新注册名，适配标准插件结构 ==》 "LighSolo"
- 新增符号 4d.Orslight ，c4d.Olight
- 新增工作流程管理，Root Space，NewProject，Make Preview(暂未开放)等

# 1004
- 修复中文排版
- Move to Top修复全局坐标错误
- Add Folder会自动添加图标tex到全局纹理路径，避免卡顿
- Add Folder加入初始相机绑定，更换为高级绑定
- Add Folder：特定文件夹会自动重命名对应层，子集为同一层（可关闭）
- 合并材质的新材质类型现在会参照第一个选择的材质

# 1005
- 新建目录更新到 v1.1
- 修复缺失文件夹bug
- 新建工程更新到 v1.2
- 增加当前工程关闭(新建工程)
- 增加当前工程移动到过程文件夹中（Process）
- 合并材质按住Ctrl会清除未使用的材质（不合并材质)

# 1006
- 更改文件夹结构为标准结构(适配插件管理器)
- 多通道EXR默认为DWAB格式,节省磁盘空间
- Octane Tag现在会自动跟随对象名称(Alt执行模式)
- 跟随custom Shortcut api更新

# 1007
- 适配PluginManager细微调整
- 全面使用2023.1.0 sdk符号,不再兼容2023之前版本
- 更新相机绑定,激活时只更改渲染视口的相机

# 1008
- 更改Folder结构，删除绑定的默认相机
- 修复Folder自动图层Bug

# 1009
- RenderPath：当渲染器为Octane时，设置info sampleing mode为without filtering（对Alpha取消AA）
- 注意：这个设置会禁用DOF和运动模糊（为了Z Depth通道更完美），如果需要，改回Distributed Rays即可
- 可编辑对象的材质默认为UVW projection（修复默认球形投射）

# 1010
- 设置渲染路径时现在会自动设置保存AE合成文件

# 1011
- 修复新建材质时选择层会冻结C4D的bug
- 选择对象时新建材质，现在新的材质球会默认在最后
- 自动化渲染路径时，现在会在文件名前添加'_'分割padding

# 1012
- 修复图层判定

# 1013
- 修复未安装redshift时的rs库导入错误
- 现在新建材质时不会自动打开材质管理器，但是如果材质管理器打开，则会定位到选择的材质
- 恢复Redshift材质，适配更快速的backlight假SSS效果
- 增加新的RS Camera对象.Alt键盘增加自定义的相机预设


res/lib/arnold_scene.py
# Copyright 2022 Autodesk, Inc.  All rights reserved.
#
# Use of this software is subject to the terms of the Autodesk license
# agreement provided at the time of installation or download, or which
# otherwise accompanies this software in either electronic or hard copy form. 

import c4d

ARNOLD_SCENE_EXPORT = 1029993

SCENE_EXPORT_FORMAT_ASS = 193450604
SCENE_EXPORT_FORMAT_USD = 193472369

SCENE_EXPORT_OBJECT_MODE_ALL = 0
SCENE_EXPORT_OBJECT_MODE_SELECTED = 1
SCENE_EXPORT_OBJECT_MODE_SELECTED_INDIVIDUALLY = 2

def Export(doc, filename, fileFormat = None, compressed = False, bbox = False, binary = True, expandProcedurals = False,
    mask = 0xFFFF, startFrame = None, endFrame = None, stepFrame = 1, objectMode = SCENE_EXPORT_OBJECT_MODE_ALL, exportObjectHierarchy = True,
    replaceWithProcedural = False):
    """
    Exports the given scene to file.

    Parameters
    ----------
    doc : c4d.documents.BaseDocument
        The scene to export.
    filename : str
        Path of the output file. If None, the file is exported next to the c4d scene file.
    fileFormat : int
        Output file format. If None, the format is guessed from the filename.
    compressed : boolean
        If true the scene is exported directly to gzip-compressed files (.ass.gz).
    bbox : boolean
        If true bounding box of the scene are added to the ASS meta data.
    binary : boolean
        If true binary encoding is used.
    expandProcedurals : boolean
        If true procedurals will be expanded before exporting the scene.
    mask : int
        Specifies which Arnold node types are included in the output. See the AtNodeEntry page of the Arnold API for the available node types.
    startFrame : int
        Specifies the first frame when exporting an animation. If None, the scene's render settings will be used.
    endFrame : int
        Specifies the last frame when exporting an animation. If None, the scene's render settings will be used.
    stepFrame : int
        Specifies the increment between frames when exporting an animation. If None, the scene's render settings will be used.
    objectMode : int
        Specifies which objects are exported (all, selected, selected individually).
    exportObjectHierarchy : boolean
        If true, exports the whole hierarchy (children) in 'selected' object mode.
    replaceWithProcedural : boolean
        If true, replaces the exported objects with an Arnold Procedural pointing to the exported file.
    """
    if doc is None:
        return

    options = c4d.BaseContainer()
    if filename is not None:
        options.SetFilename(0, filename)
    options.SetBool(1, compressed)
    options.SetBool(2, bbox)
    options.SetBool(3, binary)
    options.SetBool(4, expandProcedurals)
    options.SetInt32(5, mask)
    if startFrame is not None:
        options.SetInt32(6, startFrame)
    if endFrame is not None:
        options.SetInt32(7, endFrame)
    if stepFrame is not None:
        options.SetInt32(8, stepFrame)
    options.SetInt32(11, objectMode)
    options.SetBool(12, replaceWithProcedural)
    options.SetBool(13, exportObjectHierarchy)
    if fileFormat is not None:
        options.SetInt32(14, fileFormat)
    doc.GetSettingsInstance(c4d.DOCUMENTSETTINGS_DOCUMENT).SetContainer(ARNOLD_SCENE_EXPORT, options)
 
    c4d.CallCommand(ARNOLD_SCENE_EXPORT)



res/lib/shortcut.py
#=============================================
#                  信息 Info
#=============================================
"""
Last Update : 2022/12/01

Author: 王敦厚Go (DunHou)
Written for Maxon Cinema 4D R2023.1.0
Python version 3.9.1
Custom Cinema 4D Shortcut Functions

"""
import c4d
import typing
from typing import Optional
#=============================================
#                  介绍 Intro
#=============================================
"""
如何在Cinema 4D中查找和删除快捷方式。

在C4D内部,快捷键通过shortcut sequence绑定,通过input events侦测.


#! keySequence: 键盘输入序列, e.g., [c4d.QUALIFIER_SHIFT, '1'].

#keySequence : e.g.

    [c4d.QUALIFIER_SHIFT, c4d.QUALIFIER_ALT, "S", "T"]

对应快捷键为 SHIFT + ALT + S ~ T.

内部运算逻辑为:
    1.所有修饰键 或 在一起 (ORed) or sum
    2.支持多重连续按键 , 比如 M~S
所以 SHIFT + ALT + S ~ T 的实际数值对应为 
    [1, 4, "S", "T"]
换算为
    [(5,83),(84)]
运算为    
[  (5, 83),     # (qualifier = 1 | 4 = 5, key = ASCII_VALUE("S"))
    (84)        # (qualifier = 0        , key = ASCII_VALUE("T") ]
    
ord(): str => ASCII


#! Shortcut bc (c4d.BaseContainer): 快捷键容器结构.

#   Container Access             ID     Description
#   bc[0]                          0    第一个key stroke(修饰键序列).
#   bc[1]                          1    第一个key stroke(ASCII键值).

#   bc[10]                        10    key stroke(修饰键序列)(可选).
#   bc[11]                        11    第一个key stroke(ASCII键值)(可选).
#   [...]
#   bc[990]                      990    最大[99]key stroke(修饰键序列)(可选).
#   bc[991]                      991    最大[99]key stroke(ASCII键值)(可选).

#   bc[c4d.SHORTCUT_PLUGINID]   1000    插件ID.
#   bc[c4d.SHORTCUT_ADDRESS]    1001    指定语境(生效管理器).
#   bc[c4d.SHORTCUT_OPTIONMODE] 1002    是否打开选项.

对于多重快捷键,e.g., M ~ S, 每个index都乘10, 也就是范围[0,990], 共99个.

示例一:
    按键为 "0" .
        1."0"的ASCII编码为48.
        2.没有修饰键
        #   0: 0
        #   1: 48
        #   1000: 200000084 # pID
        #   1001: 0
        #   1002: 0    

示例二:
    按键为 "M~S" .
        1."M"的ASCII编码为77.
        2."S"的ASCII编码为83.
        3.没有修饰键
        #   0: 0
        #   1: 48
        #   1000: 200000084 # pID
        #   1001: 0
        #   1002: 0 

#! strokeData (list): 由修饰键+Key组成的元组列表

strokeData: list[tuple[int, int]]
"""

# Windows ID
OBJECT_MANAGER = 100004709
TIMELINE_MANAGER = 465001516

#=============================================
#                    Class
#=============================================
""" 
#! keySequence:  list[typing.Union[int, str]]
#! 键盘输入序列列表, e.g., [c4d.QUALIFIER_SHIFT, '1'] , [4,'2'] , [0, 'Y']

1.获取插件快捷键列表
2.获取快捷键序号(序号,不存在时返回False)
3.删除快捷键
4.获取快捷键指认的插件id和管理器id
5.添加快捷键到对应的插件id和管理器id(可选)
6.KeySequence转换StrokeData
7.检测快捷键是否已经绑定给指定插件
8.为插件添加快捷键(监测快捷键指认)
9.打印index信息
"""
class ShortCut():
    # 0.init function
    def __init__(self) -> None:
        # print("Custom shortcut library import success!")
        pass
    # _ 辅助：输入同时执行
    def check_special_input(self,QUALIFIER:int, key:str):
        bc: c4d.BaseContainer = c4d.BaseContainer()
         # Querying for a specific key with GetInputState.
         
        # Note that you must poll for upper case characters, e.g., X instead of x.
        if not c4d.gui.GetInputState(c4d.BFM_INPUT_KEYBOARD, ord (key.upper()), bc):
            raise RuntimeError("Failed to query input events.")
        # Test if the queried key is indeed being pressed.
        print (f"{bc[c4d.BFM_INPUT_VALUE] == 1 = }")
        # Test if this did co-occur with a CTRL button press.
        print (f"{bc[c4d.BFM_INPUT_QUALIFIER] == QUALIFIER = }")
    # _ 辅助：执行时输入
    def check_input_state(self):
        bc: c4d.BaseContainer = c4d.BaseContainer()
        # Querying for a specific key with GetInputState.

        if not c4d.gui.GetInputEvent(c4d.BFM_INPUT_KEYBOARD, bc):
            raise RuntimeError("Failed to query input events.")

        # Get the key which is currently be pressed as an ASCII value.
        print (f"{bc[c4d.BFM_INPUT_CHANNEL] = } ({chr (bc[c4d.BFM_INPUT_CHANNEL])})")
        # We can still read all the other things which are written into an input event container, as
        # for example a modifier key state.
        print (f"{bc[c4d.BFM_INPUT_QUALIFIER]}")  
    # 获取快捷键
    #def GetS
    # 1.获取插件快捷键元组列表
    def GetPluginShortcuts(self, pluginID: int , print_console: bool = False) -> list[list[tuple[int]]]:
        """Retrieves the shortcuts for a plugin-id.

        Args:
            pid (int): The plugin id.

        Returns:
            list[list[tuple[int]]]: The shortcut sequences for the plugin.
        """
        # Get all shortcut containers for the plugin id.
        count = c4d.gui.GetShortcutCount()
        matches = [c4d.gui.GetShortcut(i) for i in range(count)
                    if c4d.gui.GetShortcut(i)[c4d.SHORTCUT_PLUGINID] == pluginID]

        # build the shortcut data.
        result = []
        for item in matches:
            sequence = []
            for i in range(0, c4d.SHORTCUT_PLUGINID, 10):
                a, b = item[i], item[i+1]
                if isinstance(a, (int, float)):
                    sequence.append((a, b))

            if sequence != []:
                result.append(sequence)
        if print_console == True:
        # Output in console
            print("---------------")
            print("Plugin Name : {}".format(c4d.plugins.FindPlugin(pluginID).GetName()))             
            for item in result:
                for a, b in item:
                    
                    print (a, c4d.gui.Shortcut2String(a, b))
            print("---------------")
        if result == []:
            result = None
        return result
    # 2.获取快捷键全局序号(序号,不存在时返回False)
    def CheckShortcurIndex(self,keySequence: list[typing.Union[int, str]], 
                    managerId: typing.Optional[int] = None,
                    pluginId: typing.Optional[int] = None) -> bool:
        """
        Finds a shortcut index by the given #keySequence and optionally #managerId and/or #pluginId.
        False if shortcut didn't matched
            
        Args:
            keySequence: A sequence of keyboard inputs, e.g., [c4d.QUALIFIER_SHIFT, '1'].
            managerId (optional): The manager context of the shortcut to find or remove.
            pluginId (optional): The plugin ID of the plugin invoked by the shortcut.
        
        Returns:
            The success of the removal operation.

        Raises:
            RuntimeError: On illegal key symbols.
            RuntimeError: On non-existing shortcut key sequences.
        """
        # The list of key stroke modifier-key tuples.
        strokeData: list[tuple[int, int]] = []
        # A variable to OR together the qualifiers for the current key stroke.
        currentModifiers: int = 0

        #todo 获取 [strokeData]  
        #todo 转换keySequence:[c4d.QUALIFIER_SHIFT, '1'] => Output [(1, 49)]
        for key in keySequence:
            # Modifier key sequence, e.g., SHIFT + ALT + CTRL
            if isinstance(key, (int, float)):
                currentModifiers |= key
            # Character.
            elif isinstance(key, str) and len(key) == 1:
                strokeData.append((currentModifiers, ord(key.upper())))
                currentModifiers = 0
            # Errors
            else:
                raise RuntimeError(f"Found illegal key symbol: {key}")
        
        #todo Get the shortcut at #index.
        # Iterate over all shortcuts in Cinema 4D.
        for index in range(c4d.gui.GetShortcutCount()):
            
            bc: c4d.BaseContainer = c4d.gui.GetShortcut(index)

            # if #strokeData matches #bc.
            isMatch: bool = True
            for i, (qualifier, key) in enumerate(strokeData):
                idQualifier: int = i * 10 + 0
                idKey: int = i * 10 + 1
                # A qualifier + key stroke did not match, we break out.
                #? C4D中不存在此快捷键
                if bc[idQualifier] != qualifier or bc[idKey] != key:
                    isMatch = False
                    break
            
            # Something in the key sequence did not match with #strokeData, so we try the next shortcut
            # container provided by the outer loop.
            if not isMatch:
                continue
            
            # We could do here some additional tests, as shortcut key strokes do not have to be unique,
            # i.e., there could be two short-cuts "Shift + 1" bound to different manager contexts.
            if pluginId is not None and bc[c4d.SHORTCUT_PLUGINID] != pluginId:
                continue
            if managerId is not None and bc[c4d.SHORTCUT_ADDRESS] != managerId:
                continue
            
            # All tests succeeded, the shortcut at the current index should be removed, we instead just
            # return the index to make this example a bit less volatile.

            # return c4d.gui.RemoveShortcut(index)
            return index

        # All shortcuts have been traversed and no match was found, the user provided a key sequence
        # which is not a shortcut.
        return False
    # 3.删除快捷键    
    def RemoveShortcut(self,keySequence: list[typing.Union[int, str]], 
                    managerId: typing.Optional[int] = None,
                    pluginId: typing.Optional[int] = None) -> bool:
        """
        Remove Shortcut by given qualifier and key    
            
        Args:
            qualifier (int): modifier key 
            key (int): ascii code of key
        """
        index = self.CheckShortcurIndex(keySequence,managerId,pluginId)
        
        try:
            if index:
                c4d.gui.RemoveShortcut(index)
        except:
            print ("Shortcut Remove Failed")
            return False   
    # 4.获取快捷键指认的插件id和管理器id
    def FindShortcutAssign(self,keySequence: list[typing.Union[int, str]],managerId: typing.Optional[int] = None,pluginId: typing.Optional[int] = None) -> bool:
        """
        Finds a shortcut assigned plugin id and name
            
        Args:
            keySequence: A sequence of keyboard inputs, e.g., [c4d.QUALIFIER_SHIFT, '1'].
        
        Returns:
            The plugin id.
            The plugin name.
            
        Raises:
            RuntimeError: On illegal key symbols.
            RuntimeError: On non-existing shortcut key sequences.
        """
        # The list of key stroke modifier-key tuples.
        strokeData: list[tuple[int, int]] = []
        # A variable to OR together the qualifiers for the current key stroke.
        currentModifiers: int = 0

        #todo 获取 [strokeData]  
        #todo 转换keySequence:[c4d.QUALIFIER_SHIFT, '1'] => Output [(1, 49)]
        for key in keySequence:
            # Extend a modifier key sequence, e.g., SHIFT + ALT + CTRL
            if isinstance(key, (int, float)):
                currentModifiers |= key
            # A character key was found, append an input event.
            elif isinstance(key, str) and len(key) == 1:
                strokeData.append((currentModifiers, ord(key.upper())))
                currentModifiers = 0
            # Something else was found, yikes :)
            else:
                raise RuntimeError(f"Found illegal key symbol: {key}")
        
        #todo Get the shortcut at #index.
        # Now we can iterate over all shortcuts in Cinema 4D.
        for index in range(c4d.gui.GetShortcutCount()):            
            bc: c4d.BaseContainer = c4d.gui.GetShortcut(index)
            # We test if #strokeData matches #bc.
            isMatch: bool = True
            for i, (qualifier, key) in enumerate(strokeData):
                idQualifier: int = i * 10 + 0
                idKey: int = i * 10 + 1
                # A qualifier + key stroke did not match, we break out.
                #? C4D中不存在此快捷键
                if bc[idQualifier] != qualifier or bc[idKey] != key:
                    isMatch = False
                    break
            if not isMatch:
                continue

            pluginId = bc[c4d.SHORTCUT_PLUGINID]
            managerId = bc[c4d.SHORTCUT_ADDRESS]
            return pluginId,managerId
        return False
    # 5.添加快捷键到对应的插件id和管理器id(可选)     
    def AddShortCut(self,keySequence: list[typing.Union[int, str]], 
                    pluginId,
                    managerId: typing.Optional[int] = 0,
                    ) -> bool:
        """
        Add Shortcut by given qualifier and key to given ID  
            
        Args:
            qualifier (int): modifier key 
            key (int): ascii code of key
            pluginID (int): plugin ID
        """
        # The list of key stroke modifier-key tuples.
        strokeData: list[tuple[int, int]] = []
        # A variable to OR together the qualifiers for the current key stroke.
        currentModifiers: int = 0
            
        for key in keySequence:
            # Extend a modifier key sequence, e.g., SHIFT + ALT + CTRL
            if isinstance(key, (int, float)):
                currentModifiers |= key
            # A character key was found, append an input event.
            elif isinstance(key, str) and len(key) == 1:
                #strokeData.append((currentModifiers, ord(key.upper())))
                qualifier = currentModifiers
                key = ord(key.upper())
                currentModifiers = 0
            # Something else was found, yikes :)
            else:
                raise RuntimeError(f"Found illegal key symbol: {key}")
        
        for x in range(c4d.gui.GetShortcutCount()):
            shortcutBc = c4d.gui.GetShortcut(x)
            # Check if shortcut is stored in the basecontainer.        
            if shortcutBc[0] == qualifier and shortcutBc[1] == key:
                if shortcutBc[c4d.SHORTCUT_PLUGINID] == pluginId:
                    print ("Shortcut {} is already Used for Command ID: {}".format(c4d.gui.Shortcut2String(qualifier, key), shortcutBc[c4d.SHORTCUT_PLUGINID]))
                    return
            
        # Define shortcut container
        bc = c4d.BaseContainer()
        bc.SetInt32(c4d.SHORTCUT_PLUGINID, pluginId)
        bc.SetLong(c4d.SHORTCUT_ADDRESS, managerId)
        bc.SetLong(c4d.SHORTCUT_OPTIONMODE, 0)
        # User defined key
        bc.SetLong(0, qualifier)
        bc.SetLong(1, key)
        if c4d.gui.AddShortcut(bc):
            print("Shortcut Installed Susessful")
        return True
    # 6.KeySequence转换StrokeData
    #   ++》[c4d.QUALIFIER_SHIFT, '1'] => Output [(1, 49)]
    def KeySequencetoStrokeData(self,keySequence: list[typing.Union[int, str]]):
        """
        Convert Example: 
        1.[c4d.QUALIFIER_SHIFT, '1'] => Output [(1, 49)]
        2.[c4d.QUALIFIER_SHIFT, c4d.QUALIFIER_ALT, "S", "T"] => Output [(5,83),(0,84)]

        Alt : c4d.QUALIFIER_ALT = 4
        
        Ctrl : c4d.QUALIFIER_CTRL = 2
        
        Shift : c4d.QUALIFIER_SHIFT = 1

        Returns:
            _type_: _description_
        """
        strokeData: list[tuple[int, int]] = []
        # A variable to OR together the qualifiers for the current key stroke.
        currentModifiers: int = 0

        #todo 获取 [strokeData]
        #todo 转换keySequence:[c4d.QUALIFIER_SHIFT, '1'] => Output [(1, 49)]
        for key in keySequence:
            # Extend a modifier key sequence, e.g., SHIFT + ALT + CTRL
            if isinstance(key, (int, float)):
                currentModifiers |= key
            # A character key was found, append an input event.
            elif isinstance(key, str) and len(key) == 1:
                strokeData.append((currentModifiers, ord(key.upper())))
                currentModifiers = 0
            # Something else was found, yikes :)
            else:
                raise RuntimeError(f"Found illegal key symbol: {key}")
        return strokeData
    # 7.检测快捷键是否已经绑定给指定插件
    def PluginhasShortcut(self, keySequence: list[typing.Union[int, str]], pluginId:int):
        """Check if shortcut binding with given plugin
        Args:
            keySequence: A sequence of keyboard inputs, e.g., [c4d.QUALIFIER_SHIFT, '1'].            
            pluginId (optional): The plugin ID of the plugin invoked by the shortcut.

        Returns:
            bool: True if Shcrtcut with the plugin
        """
        assigned_shortcut = self.GetPluginShortcuts(pluginId)
        given_shortcut = self.KeySequencetoStrokeData(keySequence)
        if len(assigned_shortcut) == 1 and assigned_shortcut[0] == given_shortcut: # 唯一
            return True
        if len(assigned_shortcut) > 1 and given_shortcut in assigned_shortcut: # 其中之一
            return True
        if not assigned_shortcut: # 没用指认快捷键
            return False
        else:
            return False
    # 8.为插件添加快捷键(监测快捷键指认)
    def SetPluginsShortcut(self,keySequence: list[typing.Union[int, str]], pluginId:int):
        #sc_renderflow = [0, '`']
        # tip 如果插件没有指认自定义快捷键
        if self.GetPluginShortcuts(pluginId) is None :
            # tip 如果全局快捷键中没有指定快捷键
            if self.FindShortcutAssign(keySequence) == False :
                try:
                    self.AddShortCut(keySequence,pluginId)
                except:
                    raise RuntimeError("Shortcut assign Failed")
    # 9.打印index信息
    def PrintShortcutInfo(self, shortIndex:int):
        for i in shortIndex:
            bc: c4d.BaseContainer = c4d.gui.GetShortcut(i)        
            print("Index : ", i)
            print("Shortcut : ",c4d.gui.Shortcut2String(bc[0], bc[1]))
            print("PluginID : ",bc[1000])
            print("ManagerId : ",bc[1001])
            print("----------")
# exe 简化修饰键
def GetKeyMod() -> str :
    """
    Get the key str combine with the modifiers

    Returns:
        str: string with modifier key plus by + . (no space)
        1: 'Alt+Ctrl+Shift'
        2: 'Ctrl+Shift'
        3: 'Alt+Shift'
        4: 'Alt+Ctrl'
        5: 'Shift'
        6: 'Ctrl'
        7: 'Alt'
        8: 'None'
    """
    bc = c4d.BaseContainer()
    keyMod = "None" 
    if c4d.gui.GetInputState(c4d.BFM_INPUT_KEYBOARD,c4d.BFM_INPUT_CHANNEL,bc):
        if bc[c4d.BFM_INPUT_QUALIFIER] & c4d.QSHIFT:
            if bc[c4d.BFM_INPUT_QUALIFIER] & c4d.QCTRL: # Ctrl + Shift
                if bc[c4d.BFM_INPUT_QUALIFIER] & c4d.QALT: # Alt + Ctrl + Shift
                    keyMod = 'Alt+Ctrl+Shift'
                else: # Shift + Ctrl
                    keyMod = 'Ctrl+Shift'
            elif bc[c4d.BFM_INPUT_QUALIFIER] & c4d.QALT: # Alt + Shift
                keyMod = 'Alt+Shift'
            else: # Shift
                keyMod = 'Shift'
        elif bc[c4d.BFM_INPUT_QUALIFIER] & c4d.QCTRL:
            if bc[c4d.BFM_INPUT_QUALIFIER] & c4d.QALT: # Alt + Ctrl
                keyMod = 'Alt+Ctrl'
            else: # Ctrl
                keyMod = 'Ctrl'
        elif bc[c4d.BFM_INPUT_QUALIFIER] & c4d.QALT: # Alt
            keyMod = 'Alt'
        else: # No modifier
            keyMod = 'None'
        return keyMod

# # Examples
# # Shortcut Assgn when Start C4D

# 如果插件没有指认自定义快捷键，并且全局快捷键中没有指定快捷键，则为插件指认快捷键

# Import custom library to use in plugin
# import os,sys

# PluginID = 1234567 # Unique ID from www.plugincafe.com

# PLUGIN_PATH, f = os.path.split(__file__)
# Lib_Path = os.path.join(PLUGIN_PATH,"res","lib")
# sys.path.insert(0, Lib_Path)

# try:
#     import shortcut 
#     reload(shortcut)  
#     # help(shortcut)  
# finally:
#     # Remove the path we've just inserted.
#     sys.path.pop(0)

# PLUGINS Codes

# # $ Shortcut Register
# # Check shortcut add add ~ to this Plunin    
# def PluginMessage(id, data):
#     # Shortcut Assign when Start C4D
#     if id == c4d.C4DPL_PROGRAM_STARTED:
#         # tip : list[typing.Union[int, str]]
#         keySequence = [0, '`']  # ~ 
#         shortcut.ShortCut().SetPluginsShortcut(keySequence, PluginID)
#         c4d.EventAdd()
#     return False